server/backends/migration_cache/migration_cache.go

```diff
@@ -121,46 +129,106 @@ func pebbleCacheFromConfig(env environment.Env, cfg *PebbleCacheConfig) (*pebble
+func (mc *MigrationCache) Metadata(ctx context.Context, d *repb.Digest) (*interfaces.CacheMetadata, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) FindMissing(ctx context.Context, digests []*repb.Digest) ([]*repb.Digest, error) {
+func (mc *MigrationCache) FindMissing(ctx context.Context, digests []*repb.Digest) ([]*repb.Digest, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Get(ctx context.Context, d *repb.Digest) ([]byte, error) {
+func (mc *MigrationCache) GetMulti(ctx context.Context, digests []*repb.Digest) (map[*repb.Digest][]byte, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) GetMulti(ctx context.Context, digests []*repb.Digest) (map[*repb.Digest][]byte, error) {
-	return nil, status.UnimplementedError("not yet implemented")
-}
-
-func (mc MigrationCache) Set(ctx context.Context, d *repb.Digest, data []byte) error {
+func (mc *MigrationCache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error {
 	return status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error {
+func (mc *MigrationCache) Delete(ctx context.Context, d *repb.Digest) error {
 	return status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Delete(ctx context.Context, d *repb.Digest) error {
-	return status.UnimplementedError("not yet implemented")
+func (mc *MigrationCache) Reader(ctx context.Context, d *repb.Digest, offset, limit int64) (io.ReadCloser, error) {
+	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Reader(ctx context.Context, d *repb.Digest, offset, limit int64) (io.ReadCloser, error) {
+func (mc *MigrationCache) Writer(ctx context.Context, d *repb.Digest) (io.WriteCloser, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Writer(ctx context.Context, d *repb.Digest) (io.WriteCloser, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+// TODO(Maggie): Add copying logic
+func (mc *MigrationCache) Get(ctx context.Context, d *repb.Digest) ([]byte, error) {
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+	var srcBuf []byte
+
+	eg.Go(func() error {
+		srcBuf, srcErr = mc.src.Get(gctx, d)
+		return srcErr
+	})
+
+	// Double read some proportion to guarantee that data is consistent between caches
+	doubleRead := rand.Float64() <= mc.doubleReadPercentage
+	if doubleRead {
+		eg.Go(func() error {
+			_, dstErr = mc.dest.Get(gctx, d)
+			return nil // we don't care about the return error from this cache
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	if dstErr != nil {
+		// Don't log if data not found in dest cache, bc it may not have been copied yet
```

Comment:
I think this needs to be a config option. At some point we'll want to know if the source cache has items that the destination one is missing as that would be a problem. The way I see it working is that you run the cache with migrations enabled for a bit and then later you can turn off the migration and enable this verification.

Reply:
Can you think of a situation where it would be a problem if the destination cache doesn't have something that the source cache does? It would affect performance of the destination cache - are there any other risks?

Reply:
This could hide correctness issues with the destination cache. For example, if a client writes some piece of data but then the cache returns NotFound on Get for the same digest then that would indicate a correctness issue with the destination cache (ignoring eviction).

Reply:
Ack - added an option